### ISSUE TITLE
Assigning units of conductivity to colourbar scale, scale variation with regularisation parameter

### ISSUE THREAD

**Issue by AdamCoxson** [liubenyuan/pyEIT#20]
Hi, for my master's I've been using pyEIT for some work into characterising graphene, which means we'll be injecting an excitation current of 10 uA (1e-5 A) and typically measuring voltages of the mV order. We're mainly using GREIT, which is good for identifying regions of interest, but I'm struggling to determine how the actual units of conductance match the colour bar scale. Another complication is the regularisation parameter lambda, since the actual values of the scale vary for different lambda. 

When taking measurements of a resistor grid (a test stand in for graphene), we multipled all of the voltages by 1e5 to match the assumption within pyEIT that the input current is 1A. The GREIT plot scales would vary from [0.0,-0.02] for lamb=0.1 and [100, -500] for lamb=0.01. I wasn't sure how to relative the units of sieverts to the value of lambda and the scales, and hence couldn't assign units of conductivity to these scales. I could plot the L-curve for tikohonov regularisation, which would give me the 'optimal' value of lambda to use for my setup. However, this still wouldn't address the scaling and how exactly it relates to the absolute conductivity I'm measuring.

Below are the reconstructions for some oval anomaly using the opp-adj stimulation pattern for 64 electrodes around a square sample. All of the anomaly data was simulated, with a homogenous reference conductivity of 1 Sievert for the forward solution.

p = 0.2, lamb = 0.1
![image](https://user-images.githubusercontent.com/54117679/110660596-9d213a00-81bb-11eb-9481-bfcca20d0dbe.png)

p= 0.2, lamb - 0.01
![image](https://user-images.githubusercontent.com/54117679/110660617-a27e8480-81bb-11eb-89b7-92d094208d5c.png)

In the simulated case above, assuming an input current of 1 A. Would this then mean the scale on the 'true' image is in Sieverts, or Sieverts/metre, thus the blue region is -0.5 sieverts below the unit reference of 1 sievert. Hence blue = 0.5 S and yellow = 1 S. This then confuses me as to how I should relate the scale of GREIT, difference-JAC and GN to the true scale and hence obtain the absolute scaling of conductivity? Also, there must be a way of relating the scale conductivities and the lambda used to enable fixation of the units ...

Any advice on how to proceed for fixing units to this conductivity scale would be very appreciated. I'm not after a high precision of conductivity scaling, rather just the correct order of magnitudes; uS, mS, S. The only think I can think of is to use the van der Pauw method to calculate the sheet conductance of my sample and then use that value as my reference conductivity.

Many thanks,
Adam

**Comment by liubenyuan** [comment]
Hi, Nice problem! It requires to normalize the difference EIT images of different reconstruction algorithms. I can not answer this question for sure right now, and I will do some researches and follow you up soon.

**Comment by AdamCoxson** [comment]
Thank you! I've been struggling with this for a while, haha.

We found some useful papers by Alessandro Cultrera, who does the same sort of work. He used absolute solvers and that enables him to get each conductivity pixel in Sieverts. So that's why I tried the Gauss-Newton method, but even the GN colourbar scales with lambda. I'm also currently in email discussion with Alessandro, so hopefully we'll be able to work it out between us all.

Mapping the conductivity of graphene with Electrical Resistance Tomography
https://www.nature.com/articles/s41598-019-46713-8

A simple algorithm to find the L-curve corner in the regularisation of ill-posed inverse problems
https://arxiv.org/abs/1608.04571

**Comment by liubenyuan** [comment]
> Thank you! I've been struggling with this for a while, haha.
> 
> We found some useful papers by Alessandro Cultrera, who does the same sort of work. He used absolute solvers and that enables him to get each conductivity pixel in Sieverts. So that's why I tried the Gauss-Newton method, but even the GN colourbar scales with lambda. I'm also currently in email discussion with Alessandro, so hopefully we'll be able to work it out between us all.
> 
> Mapping the conductivity of graphene with Electrical Resistance Tomography
> https://www.nature.com/articles/s41598-019-46713-8
> 

1, In this paper, the authors use static EIT imaging, equation (2), where a regularization parameter lambda is used. Have you tried the static imaging method (pyeit or EIDORS)?

2, In dynamic EIT imaging, I think we need a normalization process, where a simple regression model is trained using simulation data. Is this method possible?

3, Have you tried BP method?

**Comment by AdamCoxson** [comment]
1) Yeah, so I tried the Gauss-Newton static solver within Jac.py, as shown in the 2 figures above (let me know if you cant see the figures).

2) I think I know what you mean, for example in GREIT, there is a weighting vector, W, which needs to be trained before hand? I haven't fiddled with that, I was using the default work around which I think jsut assumes w is uniform or something. I suppose I could have a deeper look. 

3) BP is giving me some trouble for some reason, B and H matrix mismatch so I get about 40 more values in the ds conductivity array. If I get it working I'll let you know if it has any difference. However, on paper_eit2016b.py demo file, the BP scale matches the true conductivity map although is off by a factor of 100.

Bit of a perplexing problem, I don't see why its so hard to assign conductivity units to the ds calculated. Seems to differ to some degree between all of the different methods ...

**Comment by liubenyuan** [comment]
I thought you were using dynamic EIT imaging. Seems like you are using static imaging.

1. The imaging matrix was calculated using **existing** meshes, for example, head phantom, lung, or circle, oval. If you know the precise shape of the problem before hand, you can generate a **perfect** mesh for this problem. Then, the normalize option (`normalize=True`) in your imaging solver may not needed, in this way, you might come to the true conductivity closer.
2. I have update the BP code yesterday, fixed a minor **normalization** problem, you can try this updated BP code. I thought BP might match, or **linearity match**, the changes in conductivities better.
3. I mean a training process for ds v.s. lambda. For example, you put the anomaly in (x, y) = (5, 5), and change its conductiviy from 1 to 1000 with 10 steps, and tune the model `f(ds, ds_true, lambda)`, then use it in your application.
4. In order to match the ds with true scale, you may need more data with a small trick. For example, combining adjacent stimulation (`el_dist=1`) with opposition stimulation (`el_dist=8`) in your code. Your imaging region is a square, which might be ill posed on the corners, may be a rounded corner version could mitigate this ill-posedness?

Hopes these could help. (I can see your figures).

**Comment by AdamCoxson** [comment]
"I thought you were using dynamic EIT imaging. Seems like you are using static imaging." - I mean, I'm using dynamic GREIT, but I've also been testing using dynamic JAC and the Static version of the Gauss Newton method to see the difference. So I'm using both.

2) I'll give the new version a shot when I get chance.
3) That's a good idea! It would at least get me closer to the true value and link to my choice of lambda parameter.
4) The problem seems to be independent of the data. I've tried opp-adj (el_dist=8), adj-adj (el_dist=1) and an adaptive electrode selection algorithm (AESA) which analyses the GREIT image iteratively so that a trained machine learning model can select a few current pairs and corresponding voltage pairs to take for the next iteration. However, I haven't tried combining opp-adj and adj-adj so I'll try that.

You've given me some good ideas to try, thank you! I've finally got my graphene sample set up so I'm focusing on taking measurements on that at the moment but I'll get back to trying to figure out the conductivity units some time next week. (I'll have to finish my master's work by 10th May, so if I haven't got it figured out by then, I'll close this thread around that date)

**Comment by AdamCoxson** [comment]
Rest of my project work took up all my time so I didn't get chance to properly look into this. I did intend to setup a script that runs over 200 anomalies and averaged over different meshes to plot the GREIT reconstruction for different p and lambda. Then normalise the respective colour bar scales to the true image scale, hopefully giving some simple factors that map the GREIT colour bar onto the true colour bar.

Unfortunately I didn't get round to this in time. I've now finished my master works so I'll close this. I think some user guide into choosing p and lambda or some developements which can map a scale onto the colour bar would prove useful. EIT has many technical parts so it can become a bit confusing for us newbies :)
